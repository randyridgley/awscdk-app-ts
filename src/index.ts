// import * as child_process from 'child_process';
import { readFileSync } from 'fs';
import * as path from 'path';
import { JsonFile, SampleDir } from 'projen';
import { TypeScriptAppProject, TypeScriptProjectOptions } from 'projen/lib/typescript';

export interface ProjCDKTypescriptOptions {

  /**
   * Package name
   */
  readonly name: string;

  /**
   * Target for synth.
   */
  readonly outdir?: string;
}

/**
 * Create a Proj CDK Typescript project.
 *
 * @pjid projcdk-app-ts
 */
export class ProjCDKTypescriptProject extends TypeScriptAppProject {
  constructor(options: ProjCDKTypescriptOptions) {

    const defaults: ProjCDKTypescriptOptions = {
      name: 'TODO',
    };

    const tsOpts: TypeScriptProjectOptions = {
      defaultReleaseBranch: 'main',
      readme: {
        filename: 'README.md',
        contents: readme(),
      },

      devDeps: [
        '@types/jest@^27.5.0',
        '@types/node@17.0.35',
        'eslint@^8.16.0',
        'jest@^27.5.1',
        'prettier@^2.7.1',
        'ts-jest@^27.1.4',
        'ts-node@^10.8.0',
        'typescript@~4.7.2',
        'eslint-config-prettier@^8.5.0',

        'aws-cdk-lib@2.40.0',
        'cdk-nag@2.18.2',
        'constructs@10.1.94',
        'cdk-monitoring-constructs@1.22.3',
        'js-yaml@4.1.0',

        '@randyridgley/awscdk-app-ts', // required for subsequent synths ('npx projen') to work!
      ],

      github: true,
      sampleCode: false,
      jest: true,
      prettier: true,
      eslint: true,
      eslintOptions: {
        prettier: true,
        dirs: ['src'],
      },

      ...defaults,
      ...options,
    };

    super(tsOpts);

    // Remove existing tasks
    this.removeTask('build');
    this.removeTask('clobber');
    this.removeTask('compile');
    this.removeTask('package');
    this.removeTask('post-compile');
    this.removeTask('post-upgrade');
    this.removeTask('pre-compile');
    this.removeTask('projen');
    this.removeTask('test-update');
    this.removeTask('upgrade');

    // Define our own tasks
    this.addTask('dependencies', { exec: this.package.installCommand, description: 'Install dependencies based on lockfile e.g. during CI' });
    this.addTask('synth', { exec: 'cdk synth --path-metadata false --version-reporting false', description: 'synth CDK stack(s)' });
    this.addTask('diff', { exec: 'cdk diff --path-metadata false --version-reporting false', description: 'diff CDK stack' });

    this.addFields({
      jest: {
        testMatch: ['<rootDir>/lib/**/*.test.ts'],
        transform: {
          '^.+\\.tsx?$': 'ts-jest',
        },
        setupFilesAfterEnv: ['./jest.setup.js'],
      },

      eslintConfig: {
        root: true,
        env: {
          node: true,
          jest: true,
        },
        extends: ['eslint-config-typescript', 'prettier'],
        parserOptions: {
          ecmaVersion: 2020,
          sourceType: 'module',
        },
        plugins: ['@typescript-eslint'],
        rules: {
          '@typescript-eslint/no-inferrable-types': 0,
          'import/no-namespace': 2,
        },
        ignorePatterns: ['**/*.js', '**/*.d.ts', 'node_modules', 'cdk.out'],
      },

    });

    new JsonFile(this, 'cdk.json', {
      obj: {
        app: 'npx ts-node lib/main.ts',
        context: {
          'aws-cdk:enableDiffNoFail': 'true',
          '@aws-cdk/core:stackRelativeExports': 'true',
        },
      },
    });

    new SampleDir(this, 'lib', {
      sourceDir: path.join(__dirname, '..', 'sample'),
    });
  }

  postSynthesize(): void {
    // const out = child_process.execSync(this.runTaskCommand(this.lintFix));
    // console.log(out.toString('utf-8'));

    console.log('Synth complete! See README.md for usage.');
  }
}

const readme = (): string => {
  return readFileSync(path.join(__dirname, '..', 'sample/README.md')).toString('utf-8');
};