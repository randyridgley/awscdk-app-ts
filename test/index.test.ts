import { Testing } from 'projen';
import { ProjCDKTypescriptProject } from '../src';

test('should match synth snapshot', () => {
  const prj = new ProjCDKTypescriptProject({ name: 'test' });
  expect(Testing.synth(prj)).toMatchSnapshot();
});