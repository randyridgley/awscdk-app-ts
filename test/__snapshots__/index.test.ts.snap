// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should match synth snapshot 1`] = `
Object {
  ".gitattributes": "# ~~ Generated by projen. To modify, edit .projenrc.js and run \\"npx projen\\".

/.gitattributes linguist-generated
/.gitignore linguist-generated
/.npmignore linguist-generated
/.projen/** linguist-generated
/.projen/deps.json linguist-generated
/.projen/files.json linguist-generated
/.projen/tasks.json linguist-generated
/cdk.json linguist-generated
/LICENSE linguist-generated
/package.json linguist-generated
/tsconfig.dev.json linguist-generated
/tsconfig.json linguist-generated
/yarn.lock linguist-generated",
  ".gitignore": "# ~~ Generated by projen. To modify, edit .projenrc.js and run \\"npx projen\\".
!/.gitattributes
!/.projen/tasks.json
!/.projen/deps.json
!/.projen/files.json
!/package.json
!/LICENSE
!/.npmignore
logs
*.log
npm-debug.log*
yarn-debug.log*
yarn-error.log*
lerna-debug.log*
report.[0-9]*.[0-9]*.[0-9]*.[0-9]*.json
pids
*.pid
*.seed
*.pid.lock
lib-cov
coverage
*.lcov
.nyc_output
build/Release
node_modules/
jspm_packages/
*.tsbuildinfo
.eslintcache
*.tgz
.yarn-integrity
.cache
!/.projenrc.js
!/test/
!/tsconfig.json
!/tsconfig.dev.json
!/src/
/lib
/dist/
!/cdk.json
",
  ".npmignore": "# ~~ Generated by projen. To modify, edit .projenrc.js and run \\"npx projen\\".
/.projen/
/test/
/tsconfig.dev.json
/src/
!/lib/
!/lib/**/*.js
!/lib/**/*.d.ts
dist
/tsconfig.json
/.github/
/.vscode/
/.idea/
/.projenrc.js
tsconfig.tsbuildinfo
",
  ".projen/deps.json": Object {
    "//": "~~ Generated by projen. To modify, edit .projenrc.js and run \\"npx projen\\".",
    "dependencies": Array [
      Object {
        "name": "@randyridgley/awscdk-app-ts",
        "type": "build",
      },
      Object {
        "name": "@randyridgley/cdk-constructs",
        "type": "build",
      },
      Object {
        "name": "@types/jest",
        "type": "build",
        "version": "^27.5.0",
      },
      Object {
        "name": "@types/node",
        "type": "build",
        "version": "^14",
      },
      Object {
        "name": "aws-cdk-lib",
        "type": "build",
        "version": "2.40.0",
      },
      Object {
        "name": "cdk-monitoring-constructs",
        "type": "build",
        "version": "1.22.3",
      },
      Object {
        "name": "cdk-nag",
        "type": "build",
        "version": "2.18.2",
      },
      Object {
        "name": "constructs",
        "type": "build",
        "version": "10.1.94",
      },
      Object {
        "name": "eslint-config-prettier",
        "type": "build",
        "version": "^8.5.0",
      },
      Object {
        "name": "eslint",
        "type": "build",
        "version": "^8.16.0",
      },
      Object {
        "name": "jest",
        "type": "build",
        "version": "^27.5.1",
      },
      Object {
        "name": "js-yaml",
        "type": "build",
        "version": "4.1.0",
      },
      Object {
        "name": "npm-check-updates",
        "type": "build",
        "version": "^15",
      },
      Object {
        "name": "prettier",
        "type": "build",
        "version": "^2.7.1",
      },
      Object {
        "name": "projen",
        "type": "build",
      },
      Object {
        "name": "ts-jest",
        "type": "build",
        "version": "^27.1.4",
      },
      Object {
        "name": "ts-node",
        "type": "build",
        "version": "^10.8.0",
      },
      Object {
        "name": "typescript",
        "type": "build",
      },
    ],
  },
  ".projen/files.json": Object {
    "//": "~~ Generated by projen. To modify, edit .projenrc.js and run \\"npx projen\\".",
    "files": Array [
      ".gitattributes",
      ".gitignore",
      ".npmignore",
      ".projen/deps.json",
      ".projen/files.json",
      ".projen/tasks.json",
      "cdk.json",
      "LICENSE",
      "tsconfig.dev.json",
      "tsconfig.json",
    ],
  },
  ".projen/tasks.json": Object {
    "//": "~~ Generated by projen. To modify, edit .projenrc.js and run \\"npx projen\\".",
    "env": Object {
      "PATH": "$(npx -c \\"node -e \\\\\\"console.log(process.env.PATH)\\\\\\"\\")",
    },
    "tasks": Object {
      "default": Object {
        "description": "Synthesize project files",
        "name": "default",
        "steps": Array [
          Object {
            "exec": "node .projenrc.js",
          },
        ],
      },
      "dependencies": Object {
        "description": "Install dependencies based on lockfile e.g. during CI",
        "name": "dependencies",
        "steps": Array [
          Object {
            "exec": "yarn install --check-files --frozen-lockfile",
          },
        ],
      },
      "diff": Object {
        "description": "diff CDK stack",
        "name": "diff",
        "steps": Array [
          Object {
            "exec": "cdk diff --path-metadata false --version-reporting false",
          },
        ],
      },
      "eject": Object {
        "description": "Remove projen from the project",
        "env": Object {
          "PROJEN_EJECTING": "true",
        },
        "name": "eject",
        "steps": Array [
          Object {
            "spawn": "default",
          },
        ],
      },
      "lint": Object {
        "description": "Lint sources using eslint",
        "name": "lint",
        "steps": Array [
          Object {
            "exec": "eslint --ext .ts --no-error-on-unmatched-pattern src/**",
          },
        ],
      },
      "synth": Object {
        "description": "synth CDK stack(s)",
        "name": "synth",
        "steps": Array [
          Object {
            "exec": "cdk synth --path-metadata false --version-reporting false",
          },
        ],
      },
      "test": Object {
        "description": "Run tests",
        "name": "test",
        "steps": Array [
          Object {
            "exec": "jest",
          },
        ],
      },
      "watch": Object {
        "description": "Watch & compile in the background",
        "name": "watch",
        "steps": Array [
          Object {
            "exec": "tsc --build -w",
          },
        ],
      },
    },
  },
  "LICENSE": "
                                 Apache License
                           Version 2.0, January 2004
                        http://www.apache.org/licenses/

   TERMS AND CONDITIONS FOR USE, REPRODUCTION, AND DISTRIBUTION

   1. Definitions.

      \\"License\\" shall mean the terms and conditions for use, reproduction,
      and distribution as defined by Sections 1 through 9 of this document.

      \\"Licensor\\" shall mean the copyright owner or entity authorized by
      the copyright owner that is granting the License.

      \\"Legal Entity\\" shall mean the union of the acting entity and all
      other entities that control, are controlled by, or are under common
      control with that entity. For the purposes of this definition,
      \\"control\\" means (i) the power, direct or indirect, to cause the
      direction or management of such entity, whether by contract or
      otherwise, or (ii) ownership of fifty percent (50%) or more of the
      outstanding shares, or (iii) beneficial ownership of such entity.

      \\"You\\" (or \\"Your\\") shall mean an individual or Legal Entity
      exercising permissions granted by this License.

      \\"Source\\" form shall mean the preferred form for making modifications,
      including but not limited to software source code, documentation
      source, and configuration files.

      \\"Object\\" form shall mean any form resulting from mechanical
      transformation or translation of a Source form, including but
      not limited to compiled object code, generated documentation,
      and conversions to other media types.

      \\"Work\\" shall mean the work of authorship, whether in Source or
      Object form, made available under the License, as indicated by a
      copyright notice that is included in or attached to the work
      (an example is provided in the Appendix below).

      \\"Derivative Works\\" shall mean any work, whether in Source or Object
      form, that is based on (or derived from) the Work and for which the
      editorial revisions, annotations, elaborations, or other modifications
      represent, as a whole, an original work of authorship. For the purposes
      of this License, Derivative Works shall not include works that remain
      separable from, or merely link (or bind by name) to the interfaces of,
      the Work and Derivative Works thereof.

      \\"Contribution\\" shall mean any work of authorship, including
      the original version of the Work and any modifications or additions
      to that Work or Derivative Works thereof, that is intentionally
      submitted to Licensor for inclusion in the Work by the copyright owner
      or by an individual or Legal Entity authorized to submit on behalf of
      the copyright owner. For the purposes of this definition, \\"submitted\\"
      means any form of electronic, verbal, or written communication sent
      to the Licensor or its representatives, including but not limited to
      communication on electronic mailing lists, source code control systems,
      and issue tracking systems that are managed by, or on behalf of, the
      Licensor for the purpose of discussing and improving the Work, but
      excluding communication that is conspicuously marked or otherwise
      designated in writing by the copyright owner as \\"Not a Contribution.\\"

      \\"Contributor\\" shall mean Licensor and any individual or Legal Entity
      on behalf of whom a Contribution has been received by Licensor and
      subsequently incorporated within the Work.

   2. Grant of Copyright License. Subject to the terms and conditions of
      this License, each Contributor hereby grants to You a perpetual,
      worldwide, non-exclusive, no-charge, royalty-free, irrevocable
      copyright license to reproduce, prepare Derivative Works of,
      publicly display, publicly perform, sublicense, and distribute the
      Work and such Derivative Works in Source or Object form.

   3. Grant of Patent License. Subject to the terms and conditions of
      this License, each Contributor hereby grants to You a perpetual,
      worldwide, non-exclusive, no-charge, royalty-free, irrevocable
      (except as stated in this section) patent license to make, have made,
      use, offer to sell, sell, import, and otherwise transfer the Work,
      where such license applies only to those patent claims licensable
      by such Contributor that are necessarily infringed by their
      Contribution(s) alone or by combination of their Contribution(s)
      with the Work to which such Contribution(s) was submitted. If You
      institute patent litigation against any entity (including a
      cross-claim or counterclaim in a lawsuit) alleging that the Work
      or a Contribution incorporated within the Work constitutes direct
      or contributory patent infringement, then any patent licenses
      granted to You under this License for that Work shall terminate
      as of the date such litigation is filed.

   4. Redistribution. You may reproduce and distribute copies of the
      Work or Derivative Works thereof in any medium, with or without
      modifications, and in Source or Object form, provided that You
      meet the following conditions:

      (a) You must give any other recipients of the Work or
          Derivative Works a copy of this License; and

      (b) You must cause any modified files to carry prominent notices
          stating that You changed the files; and

      (c) You must retain, in the Source form of any Derivative Works
          that You distribute, all copyright, patent, trademark, and
          attribution notices from the Source form of the Work,
          excluding those notices that do not pertain to any part of
          the Derivative Works; and

      (d) If the Work includes a \\"NOTICE\\" text file as part of its
          distribution, then any Derivative Works that You distribute must
          include a readable copy of the attribution notices contained
          within such NOTICE file, excluding those notices that do not
          pertain to any part of the Derivative Works, in at least one
          of the following places: within a NOTICE text file distributed
          as part of the Derivative Works; within the Source form or
          documentation, if provided along with the Derivative Works; or,
          within a display generated by the Derivative Works, if and
          wherever such third-party notices normally appear. The contents
          of the NOTICE file are for informational purposes only and
          do not modify the License. You may add Your own attribution
          notices within Derivative Works that You distribute, alongside
          or as an addendum to the NOTICE text from the Work, provided
          that such additional attribution notices cannot be construed
          as modifying the License.

      You may add Your own copyright statement to Your modifications and
      may provide additional or different license terms and conditions
      for use, reproduction, or distribution of Your modifications, or
      for any such Derivative Works as a whole, provided Your use,
      reproduction, and distribution of the Work otherwise complies with
      the conditions stated in this License.

   5. Submission of Contributions. Unless You explicitly state otherwise,
      any Contribution intentionally submitted for inclusion in the Work
      by You to the Licensor shall be under the terms and conditions of
      this License, without any additional terms or conditions.
      Notwithstanding the above, nothing herein shall supersede or modify
      the terms of any separate license agreement you may have executed
      with Licensor regarding such Contributions.

   6. Trademarks. This License does not grant permission to use the trade
      names, trademarks, service marks, or product names of the Licensor,
      except as required for reasonable and customary use in describing the
      origin of the Work and reproducing the content of the NOTICE file.

   7. Disclaimer of Warranty. Unless required by applicable law or
      agreed to in writing, Licensor provides the Work (and each
      Contributor provides its Contributions) on an \\"AS IS\\" BASIS,
      WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or
      implied, including, without limitation, any warranties or conditions
      of TITLE, NON-INFRINGEMENT, MERCHANTABILITY, or FITNESS FOR A
      PARTICULAR PURPOSE. You are solely responsible for determining the
      appropriateness of using or redistributing the Work and assume any
      risks associated with Your exercise of permissions under this License.

   8. Limitation of Liability. In no event and under no legal theory,
      whether in tort (including negligence), contract, or otherwise,
      unless required by applicable law (such as deliberate and grossly
      negligent acts) or agreed to in writing, shall any Contributor be
      liable to You for damages, including any direct, indirect, special,
      incidental, or consequential damages of any character arising as a
      result of this License or out of the use or inability to use the
      Work (including but not limited to damages for loss of goodwill,
      work stoppage, computer failure or malfunction, or any and all
      other commercial damages or losses), even if such Contributor
      has been advised of the possibility of such damages.

   9. Accepting Warranty or Additional Liability. While redistributing
      the Work or Derivative Works thereof, You may choose to offer,
      and charge a fee for, acceptance of support, warranty, indemnity,
      or other liability obligations and/or rights consistent with this
      License. However, in accepting such obligations, You may act only
      on Your own behalf and on Your sole responsibility, not on behalf
      of any other Contributor, and only if You agree to indemnify,
      defend, and hold each Contributor harmless for any liability
      incurred by, or claims asserted against, such Contributor by reason
      of your accepting any such warranty or additional liability.

   END OF TERMS AND CONDITIONS

   APPENDIX: How to apply the Apache License to your work.

      To apply the Apache License to your work, attach the following
      boilerplate notice, with the fields enclosed by brackets \\"[]\\"
      replaced with your own identifying information. (Don't include
      the brackets!)  The text should be enclosed in the appropriate
      comment syntax for the file format. We also recommend that a
      file or class name and description of purpose be included on the
      same \\"printed page\\" as the copyright notice for easier
      identification within third-party archives.

   Copyright [yyyy] [name of copyright owner]

   Licensed under the Apache License, Version 2.0 (the \\"License\\");
   you may not use this file except in compliance with the License.
   You may obtain a copy of the License at

       http://www.apache.org/licenses/LICENSE-2.0

   Unless required by applicable law or agreed to in writing, software
   distributed under the License is distributed on an \\"AS IS\\" BASIS,
   WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
   See the License for the specific language governing permissions and
   limitations under the License.
",
  "README.md": "# @randyridgley cdk Typescript app

*Generated from: github.com/randyridgley/awscdk-app-ts/*

Some example commands are:

    $ npx projen synth           // synthesise your CDK project
    $ npx projen dependencies    // install dependencies based on lockfile e.g. during CI
    $ npx projen --help          // see all available commands (test, lint, etc.)
    $ npx projen                 // update the scaffold itself

Reminder: this starter-kit uses projen (https://github.com/projen/projen).

Unlike most starter-kits, projen is not a one-off generator, and synthesized
files should not be manually edited. The only files you should edit directly
are:

* \`src/\` - your Typescript CDK files and tests
* \`environments/\` - your Yaml files for environment settings
* \`.projenrc.js\` - to update settings, e.g. to add extra dev dependencies (run
  \`npx projen\` to re-synth after any changes)

All commands should be executed via \`npx projen\` rather than targeting \`yarn\`
(or \`npm\`) directly.

## Getting started with CDK

If you are new to CDK the following resources are helpful:

* https://docs.aws.amazon.com/cdk/v2/guide/getting_started.html",
  "cdk.json": Object {
    "//": "~~ Generated by projen. To modify, edit .projenrc.js and run \\"npx projen\\".",
    "app": "npx ts-node src/main.ts",
  },
  "environments/dev.yaml": "AWSAccountID: \\"XXXXXXXXXXXX\\"
AWSProfileName: \\"default\\"
AWSProfileRegion: \\"us-east-1\\"

App: \\"DevApplication\\"
Environment: \\"dev\\"
Version: \\"0.0.0\\"
Build: \\"0\\"

Parameters:
  TestParameter: \\"TestFooBar\\"
",
  "environments/prod.yaml": "AWSAccountID: \\"XXXXXXXXXXXX\\"
AWSProfileName: \\"default\\"
AWSProfileRegion: \\"us-east-1\\"

App: \\"ProdApplication\\"
Environment: \\"prod\\"
Version: \\"0.0.0\\"
Build: \\"0\\"

Parameters:
  TestParameter: \\"ProdFooBar\\"
",
  "package.json": Object {
    "//": "~~ Generated by projen. To modify, edit .projenrc.js and run \\"npx projen\\".",
    "devDependencies": Object {
      "@randyridgley/awscdk-app-ts": "*",
      "@randyridgley/cdk-constructs": "*",
      "@types/jest": "^27.5.0",
      "@types/node": "^14",
      "aws-cdk-lib": "2.40.0",
      "cdk-monitoring-constructs": "1.22.3",
      "cdk-nag": "2.18.2",
      "constructs": "10.1.94",
      "eslint": "^8.16.0",
      "eslint-config-prettier": "^8.5.0",
      "jest": "^27.5.1",
      "js-yaml": "4.1.0",
      "npm-check-updates": "^15",
      "prettier": "^2.7.1",
      "projen": "*",
      "ts-jest": "^27.1.4",
      "ts-node": "^10.8.0",
      "typescript": "*",
    },
    "eslintConfig": Object {
      "env": Object {
        "jest": true,
        "node": true,
      },
      "extends": Array [
        "eslint-config-typescript",
        "prettier",
      ],
      "ignorePatterns": Array [
        "**/*.js",
        "**/*.d.ts",
        "node_modules",
        "cdk.out",
      ],
      "parserOptions": Object {
        "ecmaVersion": 2020,
        "sourceType": "module",
      },
      "plugins": Array [
        "@typescript-eslint",
      ],
      "root": true,
      "rules": Object {
        "@typescript-eslint/no-inferrable-types": 0,
        "import/no-namespace": 2,
      },
    },
    "jest": Object {
      "setupFilesAfterEnv": Array [
        "./jest.setup.js",
      ],
      "testMatch": Array [
        "<rootDir>/src/**/*.test.ts",
      ],
      "transform": Object {
        "^.+\\\\.tsx?$": "ts-jest",
      },
    },
    "license": "Apache-2.0",
    "name": "test",
    "resolutions": Object {
      "@types/responselike": "1.0.0",
      "got": "12.3.1",
    },
    "scripts": Object {
      "default": "npx projen default",
      "dependencies": "npx projen dependencies",
      "diff": "npx projen diff",
      "eject": "npx projen eject",
      "lint": "npx projen lint",
      "projen": "npx projen",
      "synth": "npx projen synth",
      "test": "npx projen test",
      "watch": "npx projen watch",
    },
    "version": "0.0.0",
  },
  "src/environment/build-config.ts": "export interface BuildConfig {
    readonly AWSAccountID : string;
    readonly AWSProfileName : string;
    readonly AWSProfileRegion : string;

    readonly App : string;
    readonly Environment : string;
    readonly Version : string;
    readonly Build : string;

    readonly Parameters: Parameters;
}

export interface Parameters {
    readonly TestParameter?: string;
}",
  "src/main.ts": "import path from 'path';
import * as fs from 'fs'
import { App, Aspects } from 'aws-cdk-lib';
import { AwsSolutionsChecks } from 'cdk-nag';
import { DefaultStack } from './stacks/default-stack';
import { BuildConfig } from './environment/build-config';
const yaml = require('js-yaml');

const app = new App();

function ensureString(object: { [name: string]: any }, propName: string ): string {
    if(!object[propName] || object[propName].trim().length === 0)
        throw new Error(propName +\\" does not exist or is empty\\");

    return object[propName];
}

function getConfig() {
    let env = app.node.tryGetContext('config');
    if (!env)
        throw new Error(\\"Context variable missing on CDK command. Pass in as \`-c config=XXX\`\\");

    let unparsedEnv = yaml.load(fs.readFileSync(path.resolve(\\"./environments/\\"+env+\\".yaml\\"), \\"utf8\\"));

    let buildConfig: BuildConfig = {
        AWSAccountID: ensureString(unparsedEnv, 'AWSAccountID'),
        AWSProfileName: ensureString(unparsedEnv, 'AWSProfileName'),
        AWSProfileRegion: ensureString(unparsedEnv, 'AWSProfileRegion'),

        App: ensureString(unparsedEnv, 'App'),
        Version: ensureString(unparsedEnv, 'Version'),
        Environment: ensureString(unparsedEnv, 'Environment'),
        Build: ensureString(unparsedEnv, 'Build'),

        Parameters: {
            TestParameter: ensureString(unparsedEnv['Parameters'], 'TestParameter'),
        }
    };

    return buildConfig;
}

let buildConfig: BuildConfig = getConfig();

let defaultStackName = buildConfig.App + \\"-\\" + buildConfig.Environment + \\"-default\\";
const defaultStack = new DefaultStack(app, defaultStackName, {
  env: {
      region: buildConfig.AWSProfileRegion,
      account: buildConfig.AWSAccountID
  },
}, buildConfig);

Aspects.of(defaultStack).add(new AwsSolutionsChecks({ verbose: true }));

app.synth();",
  "src/stacks/base-stack.ts": "import { NestedStack, NestedStackProps, Stack, StackProps, Tags } from 'aws-cdk-lib';
import { Construct } from 'constructs';
import { BuildConfig } from '../environment/build-config';

/**
 * Default tags for all stacks
 */
export class BaseStack extends Stack {
  constructor(scope: Construct, id: string, props: StackProps, buildConfig: BuildConfig) {
    super(scope, id, props);

    //Tagging
    Tags.of(this).add('Environment', \`\${buildConfig.Environment}\`);
    Tags.of(this).add('Application', \`\${buildConfig.App}\`);
    Tags.of(this).add('Version', \`\${buildConfig.Version}\`);
    Tags.of(this).add('Stack', this.stackName);
  }
}

/**
 * Add tags to all nested stacks
 */
export class NestedBaseStack extends NestedStack {
  constructor(scope: Construct, id: string, props?: NestedStackProps) {
    super(scope, id, props);

    if (this.nestedStackParent?.tags.hasTags) {
      Object.entries(this.nestedStackParent.tags.tagValues()).forEach(([key, value]) => {
        this.tags.setTag(key, value);
      });
    }
  }
}
",
  "src/stacks/default-stack.ts": "import { StackProps } from 'aws-cdk-lib';
import { Construct } from 'constructs';
import { BuildConfig } from '../environment/build-config';
import { BaseStack } from './base-stack';

export interface DefaultStackProps extends StackProps {
  
}

export class DefaultStack extends BaseStack {
  private readonly buildConfig: BuildConfig;

  constructor(scope: Construct, id: string, props: DefaultStackProps, buildConfig: BuildConfig) {
    super(scope, id, props, buildConfig);
    this.buildConfig = buildConfig;
    console.log(this.buildConfig.Parameters.TestParameter);
  }
}
",
  "tsconfig.dev.json": Object {
    "//": "~~ Generated by projen. To modify, edit .projenrc.js and run \\"npx projen\\".",
    "compilerOptions": Object {
      "alwaysStrict": true,
      "declaration": true,
      "esModuleInterop": true,
      "experimentalDecorators": true,
      "inlineSourceMap": true,
      "inlineSources": true,
      "lib": Array [
        "es2019",
      ],
      "module": "CommonJS",
      "noEmitOnError": false,
      "noFallthroughCasesInSwitch": true,
      "noImplicitAny": true,
      "noImplicitReturns": true,
      "noImplicitThis": true,
      "noUnusedLocals": true,
      "noUnusedParameters": true,
      "resolveJsonModule": true,
      "strict": true,
      "strictNullChecks": true,
      "strictPropertyInitialization": true,
      "stripInternal": true,
      "target": "ES2019",
    },
    "exclude": Array [
      "node_modules",
    ],
    "include": Array [
      ".projenrc.js",
      "src/**/*.ts",
      "test/**/*.ts",
    ],
  },
  "tsconfig.json": Object {
    "//": "~~ Generated by projen. To modify, edit .projenrc.js and run \\"npx projen\\".",
    "compilerOptions": Object {
      "alwaysStrict": true,
      "declaration": true,
      "esModuleInterop": true,
      "experimentalDecorators": true,
      "inlineSourceMap": true,
      "inlineSources": true,
      "lib": Array [
        "es2019",
      ],
      "module": "CommonJS",
      "noEmitOnError": false,
      "noFallthroughCasesInSwitch": true,
      "noImplicitAny": true,
      "noImplicitReturns": true,
      "noImplicitThis": true,
      "noUnusedLocals": true,
      "noUnusedParameters": true,
      "outDir": "lib",
      "resolveJsonModule": true,
      "rootDir": "src",
      "strict": true,
      "strictNullChecks": true,
      "strictPropertyInitialization": true,
      "stripInternal": true,
      "target": "ES2019",
    },
    "exclude": Array [],
    "include": Array [
      "src/**/*.ts",
    ],
  },
}
`;
